p4-16/spec/P4-16-spec.mdk

```diff
@@ -7097,16 +7099,157 @@ annotationBody
     | annotationBody annotationToken
 ~ End P4Grammar
 
-Annotations may impose additional structure on their bodies, and are not
-confined to the P4 language. For example, the P4Runtime specification defines a
-`@pkginfo` annotation that takes key-value pairs, which are not found in the
-main P4 language.
+Unstructured annotations may impose additional structure on their bodies, and
+are not confined to the P4 language. For example, the P4Runtime specification
+defines a `@pkginfo` annotation that takes key-value pairs, which are not found
+in the main P4 language.
+
+## Bodies of Structured Annotations
+
+Unlike unstructured annotations, structured annotations use square brackets
+`[...]` and have a specific format. They are commonly used to declare custom
+metadata, consisting of expression lists or key-value lists but not both (except
+that the value of a `kvPair` may itelf be an `expressionList`). An
+`expressionList` may be empty or contain a comma-separated list of members. A
+`kvList` consists of one or more `kvPair`s, each consisting of a key and a
+value; the value is an `expressionList`. 
+
+To avoid ambiguity, an empty `stucturedAnnotationBody` is by definition
+considered to contain an empty `expressionList`. Furthermore, the value of a `kvPair` may
+be the empty `expressionList`, in which case the pair is considered to
+contain a key without a value.
+
+The `expression`s within an `expressionList` must evaluate at compile-time to
```

Comment:
This is actually not true for annotations such as `@length`; in that annotation the expression actually refers to some structure field members and leads to code being generated to evaluate the expression dynamically.

I notice that the `@length` annotation is actually not documented in the P4-16 spec; it is used to represent p4-14 varbit field length information. I wonder whether we should document it. 

The point is that we do not mandate in the spec when and how the expression that is an argument to the annotation is evaluated. It does not have to be a compile-time evaluated expression.

Reply:
Are you suggesting that these restrictions should appear in the P4Runtime spec instead?